torchqdynamics/ode_forward_qsolver.py

```diff
@@ -0,0 +1,167 @@
+from .progress_bar import tqdm
+from .qsolver import QSolver
+from .solver_options import Dopri45, ODEAdaptiveStep, ODEFixedStep
+
+
+class ODEForwardQSolver(QSolver):
+    @abstractmethod
+    def forward(self, t: float, y: Tensor) -> Tensor:
+        """Iterate the quantum state forward.
+
+        Args:
+            t: Time.
+            y: Quantum state, of shape `(..., m, n)`.
+
+        Returns:
+            Tensor of shape `(..., m, n)`.
+        """
+        pass
+
+    def run(self):
+        """Integrate a quantum ODE starting from an initial state.
+
+        The ODE is solved from time `t=0.0` to `t=t_save[-1]`.
+        """
+        # dispatch to appropriate odeint subroutine
+        if self.gradient_alg is None:
+            self._odeint_inplace()
+        elif self.gradient_alg == 'autograd':
+            self._odeint_main()
+
+    def _odeint_inplace(self):
+        """Integrate a quantum ODE with an in-place solver.
+
+        Simple solution for now so torch does not store gradients.
+        TODO Implement a genuine in-place solver.
+        """
+        with torch.no_grad():
+            self._odeint_main()
+
+    def _odeint_main(self):
+        """Dispatch the ODE integration to fixed or adaptive time step subroutines."""
+        if isinstance(self.options, ODEFixedStep):
+            self._fixed_odeint()
+        elif isinstance(self.options, ODEAdaptiveStep):
+            self._adaptive_odeint()
+
+    def _fixed_odeint(self):
+        """Integrate a quantum ODE with a fixed time step solver.
+
+        Note:
+            The solver times are defined using `torch.linspace` which ensures that the
+            overall solution is evolved from the user-defined time (up to an error of
+            `rtol=1e-5`). However, this may induce a small mismatch between the time
+            step inside `qsolver` and the time step inside the iteration loop. A small
+            error can thus buildup throughout the ODE integration. TODO Fix this.
+        """
+        # get time step
+        dt = self.options.dt
+
+        # assert that `t_save` values are multiples of `dt`
+        if not torch.allclose(torch.round(self.t_save / dt), self.t_save / dt):
+            raise ValueError(
+                'For fixed time step solvers, every value of `t_save` must be a'
+                ' multiple of the time step `dt`.'
+            )
+
+        # define time values
+        num_times = torch.round(self.t_save[-1] / dt).int() + 1
+        times = torch.linspace(0.0, self.t_save[-1], num_times)
+
+        # run the ode routine
+        y = self.y0
+        for t in tqdm(times[:-1], disable=not self.options.verbose):
+            # save solution
+            if t >= self.next_tsave():
+                self.save(y)
+
+            # iterate solution
+            y = self.forward(t, y)
+
+        # save final time step (`t` goes `0.0` to `t_save[-1]` excluded)
+        self.save(y)
+        self.save_final(y)
```

Comment:
Kind of strange to have `self.save(y)` twice here. It looks like you're saving the state twice. IMO a better logic would be to have only `self.save_final(y)` and within the corresponding function, to check for the `self.options.save_states` boolean.